plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java

```diff
@@ -0,0 +1,331 @@
+import static java.lang.Boolean.parseBoolean;
+import static java.lang.invoke.MethodHandles.lookup;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+import static org.apache.hadoop.hive.metastore.TableType.MANAGED_TABLE;
+import static org.apache.iceberg.BaseMetastoreTableOperations.ICEBERG_TABLE_TYPE_VALUE;
+import static org.apache.iceberg.BaseMetastoreTableOperations.METADATA_LOCATION_PROP;
+import static org.apache.iceberg.BaseMetastoreTableOperations.TABLE_TYPE_PROP;
+import static org.apache.iceberg.TableProperties.DEFAULT_NAME_MAPPING;
+import static org.apache.iceberg.TableProperties.FORMAT_VERSION;
+import static org.apache.iceberg.mapping.NameMappingParser.toJson;
+
+public class MigrateProcedure
+        implements Provider<Procedure>
+{
+    private static final Logger log = Logger.get(MigrateProcedure.class);
+
+    public static final String PROVIDER_PROPERTY_KEY = "provider";
+    public static final String PROVIDER_PROPERTY_VALUE = "iceberg";
+    private static final List<String> EXCLUDED_PROPERTIES = ImmutableList.of("path", "transient_lastDdlTime", "serialization.format");
+
+    // TODO: Set row count as 0L because DataFile disallows null. This is same as Spark implementation, but we should store the actual row count.
+    private static final Metrics EMPTY_METRICS = new Metrics(0L, null, null, null, null);
+
+    private final TrinoCatalogFactory catalogFactory;
+    private final HiveMetastoreFactory metastoreFactory;
+    private final TrinoFileSystemFactory fileSystemFactory;
+    private final TypeManager typeManager;
+    private final int formatVersion;
+    private final boolean isUsingSystemSecurity;
+
+    private static final MethodHandle MIGRATE;
+
+    static {
+        try {
+            MIGRATE = lookup().unreflect(MigrateProcedure.class.getMethod("migrate", ConnectorSession.class, String.class, String.class));
+        }
+        catch (ReflectiveOperationException e) {
+            throw new AssertionError(e);
+        }
+    }
+
+    @Inject
+    public MigrateProcedure(
+            TrinoCatalogFactory catalogFactory,
+            HiveMetastoreFactory metastoreFactory,
+            TrinoFileSystemFactory fileSystemFactory,
+            TypeManager typeManager,
+            IcebergConfig icebergConfig,
+            IcebergSecurityConfig securityConfig)
+    {
+        this.catalogFactory = requireNonNull(catalogFactory, "catalogFactory is null");
+        this.metastoreFactory = requireNonNull(metastoreFactory, "metastoreFactory is null");
+        this.fileSystemFactory = requireNonNull(fileSystemFactory, "fileSystemFactory is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.formatVersion = requireNonNull(icebergConfig, "icebergConfig is null").getFormatVersion();
+        this.isUsingSystemSecurity = requireNonNull(securityConfig, "securityConfig is null").getSecuritySystem() == SYSTEM;
```

Comment:
`requireNonNull` is redundant here and for `icebergConfig` above too (https://github.com/trinodb/trino/pull/13940)